feat(aci): add DetectorWorkflow index endpoint

ameliahsu opened the pull request:
add GET/POST/DELETE for DetectorWorkflow `/organizations/:id_or_slug/detector-workflows`

ameliahsu committed:
feat(aci): add DetectorWorkflow endpoint

ameliahsu committed:
add DetectorWorkflow.id

ameliahsu commented on src/sentry/workflow_engine/endpoints/organization_detector_workflow_index.py:
```diff
@@ -0,0 +1,140 @@
+        "POST": ApiPublishStatus.EXPERIMENTAL,
+        "DELETE": ApiPublishStatus.EXPERIMENTAL,
+    }
+    owner = ApiOwner.ISSUES
```
realizing we probably need to add `permission_classes` here, but I'm not sure which permissions to use?

ameliahsu committed:
add OrganizationPermission

ameliahsu commented on src/sentry/workflow_engine/endpoints/organization_detector_workflow_index.py:
```diff
@@ -0,0 +1,142 @@
+        "DELETE": ApiPublishStatus.EXPERIMENTAL,
+    }
+    owner = ApiOwner.ISSUES
+    permission_classes = OrganizationPermission
```
does a user need project permission to the detectors in order to connect them?

ameliahsu replied:
i think having the `project_detector` URL pattern makes sense for POST since that's connecting an single detector to a single workflow, but if GET and DELETE are meant for bulk actions then this wouldn't work since workflows can be connected to detectors from multiple projects 🤔  this is tricky

ameliahsu replied:
Update: decided that creation/deletion of DetectorWorkflow is determined by detector permissions

ameliahsu committed:
fix type

ameliahsu committed:
extra tests yippee

ameliahsu committed:
add audit logging

ceorourke commented on tests/sentry/workflow_engine/endpoints/test_organization_detector_workflow_index.py:
```diff
@@ -0,0 +1,220 @@
+            "workflowId": str(self.unconnected_workflow.id),
+        }
+
+    def test_audit_entry(self):
```
Here's an alternate way to add in checking audit log entries. Mock it like this: https://github.com/getsentry/sentry/blob/master/tests/sentry/workflow_engine/endpoints/test_project_detector_index.py#L121 and then check it here: https://github.com/getsentry/sentry/blob/master/tests/sentry/workflow_engine/endpoints/test_project_detector_index.py#L168-L175

ameliahsu replied:
hmm i just tried this but i don't see a way to check for multiple audit log entries this way? i need to assert that all DetectorWorkflows have an audit log in the case of bulk deletion

ceorourke replied:
There are several options for mock assertion (docs [here](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.assert_called)) so you don't just need to assert it's called once with, maybe you could loop over the `DetectorWorkflow` objects and `assert_called_with`

ameliahsu replied:
i realize that i messed something up the first time, i'll try again ty 🫡

ameliahsu commented on src/sentry/workflow_engine/endpoints/organization_detector_workflow_index.py:
```diff
@@ -0,0 +1,161 @@
+
+        if not detector_id and not workflow_id:
+            raise serializers.ValidationError(
+                {"detail": "detector_id or workflow_id must be provided."}
```
we only need one, since we want to be able to bulk delete DetectorWorkflows (if only `detector_id` is provided, we delete all DetectorWorkflows with the specified detector)

ceorourke replied:
Got it, could you update the docstring to be more descriptive about the inputs and behaviors?

ceorourke replied:
Wait a second, 107 is checking that _neither_ were passed, but then the error message is saying you only passed one but not the other. So should 107 be an _or_ check? Can you add a test case for this too?

ameliahsu replied:
Maybe the wording is a bit confusing, but the error is saying that detector_id or workflow_id (one or the other, or both) must be provided, since none were passed. I updated the docstring to explain!

ameliahsu commented on src/sentry/workflow_engine/endpoints/organization_detector_workflow_index.py:
```diff
@@ -0,0 +1,161 @@
+
+        queryset = DetectorWorkflow.objects.filter(workflow__organization=organization)
+
+        if workflow_id:
```
this relates to the comment above, but we check for workflow_id and detector_id since only one needs to be provided

ameliahsu commented on src/sentry/workflow_engine/endpoints/organization_detector_workflow_index.py:
```diff
@@ -0,0 +1,161 @@
+        if not queryset:
+            return Response(status=status.HTTP_404_NOT_FOUND)
+
+        for detector_workflow in queryset:
```
if only detector_id or workflow_id is passed, we want to delete all DetectorWorkflows associated with the specified detector/workflow. this is tested here, where we delete all DetectorWorkflows where `detector.id = self.detector_1.id`

https://github.com/getsentry/sentry/blob/ed832c0f08b3375dbcd9edd75af681428309837d/tests/sentry/workflow_engine/endpoints/test_organization_detector_workflow_index.py#L184-L193

ameliahsu commented on src/sentry/workflow_engine/endpoints/serializers.py:
```diff
@@ -282,3 +283,20 @@ def serialize(self, obj: Workflow, attrs: Mapping[str, Any], user, **kwargs) ->
         }
+
+
+class DetectorWorkflowResponse(TypedDict):
```
oh oops ty i meant to type the response of the serializer below

ameliahsu committed:
fixes

ameliahsu committed:
more/better tests

ceorourke reviewed (approved):
looks good, thanks for the updates, just need to rebase

ameliahsu committed:
nit

ameliahsu committed:
feat(aci): detector permission checks for DetectorWorkflow endpoint (#88626)

detectors should have similar permissions to alerts today, with
additional restrictions on default detectors created by Sentry

- Sentry created (default) detectors
- Organization-level owner, manager (users with `org:write` or
`org:admin`)
- Team-level admins for projects where they are team admins (those who
have `project:write` on the project)
- User created (custom) detectors
- Anyone on the team can create/edit if “Let members create alerts” is
enabled (in this case all users have `alert:write` on projects)
- Otherwise if “Let members create alerts” is disabled, only
organization-level owner, manager, or team-level admin roles can
create/edit detectors
- These permissions should be restricted to only projects that members
are associated with in the case that "Open team membership" is disabled
